src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -288,8 +288,42 @@ public static int FromLowerChar(int c)
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static bool IsHexChar(int c)
+        public static unsafe bool IsHexChar(int c)
         {
+            if (sizeof(IntPtr) == 8)
+            {
+                // Same logic as BitHelper.HasLookupFlag in Microsoft.Toolkit.HighPerformance.
+                // First, the input value is scaled down by '0' (48), which is the offset used
+                // to populate the bitmask. This avoids wasting 48 bits set to 0 at the start of
+                // the mask, and lets all the necessary range (from '0' to 'f' to fit in 64 bits).
+                // We then do a range check (the clt.un comparison will push 0/1, so the bool will
+                // be normalized) and use the result to compute a bitwise flag of either 0xFFFFFFFF
+                // if the input is accepted, or all 0 otherwise. The target bit is then extracted,
+                // (the mask has a 1 bit for "0123456789ABCDEFabcdef" chars left shifted by '0')
+                // and this value is combined with the previous mask. This is done so that
+                // if the shift was performed with a value that was too high, which has an
+                // undefined behavior and could produce a non-0 value, the mask will reset
+                // the final value anyway. This result is then unchecked-cast to a byte (as
+                // it is guaranteed to always be either 1 or 0), and than to a bool to return.
+                // We only use this logic on 64-bit systems, as using 64 bit values would otherwise
+                // be much slower than just using the lookup table anyway (no hardware support).
+                // This code path, when used, has no branches and doesn't depend on cache hits,
+                // so it's faster and does not vary in speed depending on input data distribution.
+                // It also gets JITted to just a constant when the input is a constant as well.
+                int i = c - '0';
+                bool isInRange = (uint)i < 64u;
+                byte byteFlag = *(byte*)&isInRange;
+                int
+                    negativeFlag = byteFlag - 1,
+                    mask = ~negativeFlag,
+                    shift = unchecked((int)((35465847073801215UL >> i) & 1)),
+                    and = shift & mask;
+                byte result = unchecked((byte)and);
+                bool valid = *(bool*)&result;
```

Comment:
We considered similar tricks to this a while back and those tricks were rejected. See https://github.com/dotnet/coreclr/pull/16138 for one example.

The tl;dr was "somebody's probably branching on the result of this method, so it's better if the final statement is an actual comparison that can be folded into the caller's branch condition."

Reply:
That makes sense, I was thinking the last return expression could be changed to that due to callers branching 🙂
Question though: given the input is guaranteed to be either `1` or `0` here due to the previous logic, we can just test `and != 0` here, right? As in, we can skip that `& 1` since that wouldn't affect the actual semantics in this case, no?

So at the end of the day:
```csharp
return and != 0 ? true : false;
```
Should work?

Reply:
Yup, that should work!